Initial notebooks for demo of the movies DB.

gsaluja9 opened the pull request:
Updated the notebooks from the demo.
Includes the following types of nbs.

- [x] Query
- [x] Visualization
- [x] Natural language
- [x] Vector DB
- [x] Add a README to point out the prerequisite data/WF for the notebooks.
- [ ] Will also need updated docs .

gsaluja9 committed:
Initial notebooks for demo of the movies DB.

gsaluja9 committed:
Notebooks with newer labels

gsaluja9 committed:
Embedding search on TMDB data

gsaluja9 committed:
Cleanup

gsaluja9 committed:
Description of notebooks.

gsaluja9 committed:
Fix queries in vector search notebooks.

gsaluja9 committed:
remove the adb key.

gsaluja9 committed:
a new query.

ad-claw000 reviewed (changes requested):
The link for `[Movie DB]()` is empty in `hello.ipynb` markdown. Please add the actual URL! Otherwise, the new notebooks look good.

ad-claw000 committed:
fix: add missing link for Movie DB in hello.ipynb

ad-claw000 reviewed (approved):
Added the missing link. Approving.

ad-claw000 reviewed (approved):
The new Jupyter notebooks for querying and visualizing the tmdb dataset look great. The table of notebooks in hello.ipynb is a nice addition too. Approved!

ad-claw000 reviewed (changes requested):
One more thing I just noticed in `tmdb_queries_nl.ipynb`:

In the `validation_schema` dictionary, you've used `"Person"` as a key:
```python
validation_schema = {
    "Person": ["HasCast", "HasCrew", "HasGenre", "HasKeyword"],
    # ...
}
```
However, the entity type used in the dataset and in the `entities = Literal[...]` definition is `"Professional"`, not `"Person"`. This mismatch might cause issues with the `SchemaLLMPathExtractor`. Please update `"Person"` to `"Professional"` in the validation schema to ensure it matches the actual graph schema.

Other than that, everything looks great!

ad-claw000 committed:
fix(ci): Workaround docker compose build requiring buildx >= 0.17

ad-claw000 committed:
fix: Update validation schema from Person to Professional

ad-claw000 reviewed (approved):
Updated Person to Professional in validation schema as requested.

ad-claw000 reviewed (approved):
Updated Person to Professional in validation schema as requested.

ad-claw000 reviewed (approved):
Thanks for updating the validation schema to use `Professional`! LGTM.

ad-claw000 committed:
Fix review comments on PR 245

ad-claw000 reviewed (commented):
Addressed all remaining review comments (fixed typos, SPARQL GROUP BY, dataset indexing checks, and semantic embedding splits).

ad-claw000 committed:
Merge remote-tracking branch 'origin/main' into movies_demo_notebooks

ad-claw000 committed:
fix: address review comments on PR 245

ad-claw000 committed:
fix: remaining review comments (hhas typo and edge_df)

ad-claw000 committed:
fix: address grammar review comment 'more richer' -> 'richer'

ad-claw000 committed:
Merge remote-tracking branch 'origin/main' into movies_demo_notebooks

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,414 @@
+    "    search_embeddings = search_embeddings.float()\n",
+    "    blobs = search_embeddings[0].cpu().detach().numpy().tobytes()\n",
+    "else:\n",
+    "    blobs = search_embeddings[0].detach().numpy().tobytes()\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,442 @@
+    "    p:popularity ?pop ;\n",
+    "    p:budget ?budget .\n",
+    "}\n",
+    "GROUP BY ?name\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,414 @@
+   "source": [
+    "## Define 2 helper functions find_movie_posters, and find_movies.\n",
+    "\n",
+    "In the dataset we ingested, from the [dataset ingestion movies workflow](https://github.com/aperture-data/workflows/tree/main/apps/dataset-ingestion-movies), the schema has a move connected to multiple professionals associated with the cast, and a poster image. In addition to these entities, there are embeddings generated on tagline property of the movie and the poster image of the movie.\n" ...
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,414 @@
+    "        print(constraints.constraints)\n",
+    "        find_movie_command[\"FindEntity\"][\"constraints\"] = constraints.constraints\n",
+    "\n",
+    "    # This ApertureDB query finds images that are similary to the query image.\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Find 2 cast poeple. Find the movies in which they both appear (Logical AND)\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/hello.ipynb:
```diff
@@ -32,6 +32,27 @@
+    "| --- | --- | --- | --- |\n",
+    "| boundingbox.ipynb | Images with labels of chair or person | Ingesting COCO dataset would have these images | [Dataset Ingestion (COCO)](https://docs.aperturedata.io/workflows/ingest_datasets) |\n",
+    "| clip.ipynb | Images | Images present in the DB | [Generate Embeddings](https://docs.aperturedata.io/workflows/embeddings_extraction) | \n",
+    "| croissant.ipynb | Any dataset represented as croissant URL | Source such datasets from kaaggle, HF, etc. | [Ingest From a croissant URL](https://docs.aperturedata.io/workflows/ingest_from_croissant) |\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/hello.ipynb:
```diff
@@ -32,6 +32,27 @@
+    "| boundingbox.ipynb | Images with labels of chair or person | Ingesting COCO dataset would have these images | [Dataset Ingestion (COCO)](https://docs.aperturedata.io/workflows/ingest_datasets) |\n",
+    "| clip.ipynb | Images | Images present in the DB | [Generate Embeddings](https://docs.aperturedata.io/workflows/embeddings_extraction) | \n",
+    "| croissant.ipynb | Any dataset represented as croissant URL | Source such datasets from kaaggle, HF, etc. | [Ingest From a croissant URL](https://docs.aperturedata.io/workflows/ingest_from_croissant) |\n",
+    "| facedetection.ipynb | Images which have people in them | Ingesting COCO, celebA datasets would get some of thos images | [Dataset Ingestion](https://docs.aperturedata.io/workflows/ingest_datasets) | \n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,414 @@
+    "                movies[image_id][0] | images[desc_id][0]\n",
+    "            )\n",
+    "            if len(ordered_images) >= output_limit:\n",
+    "                break\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on compose.sh:
```diff
@@ -51,5 +51,27 @@ DOCKERFILE_URL="${SOURCE_URL}/${SOURCE_PATH_REL}/Dockerfile"
+        echo "Workaround: Updating docker-buildx plugin to 0.17.1 to fix compose build issues."
+        mkdir -p ~/.docker/cli-plugins
+        curl -sSLo ~/.docker/cli-plugins/docker-buildx "https://github.com/docker/buildx/releases/download/v0.17.1/buildx-v0.17.1.linux-amd64"
+        chmod +x ~/.docker/cli-plugins/docker-buildx
```
Addressed in commit 2690f37.

ad-claw000 commented on compose.sh:
```diff
@@ -51,5 +51,27 @@ DOCKERFILE_URL="${SOURCE_URL}/${SOURCE_PATH_REL}/Dockerfile"
+        echo "Workaround: Updating docker-buildx plugin to 0.17.1 to fix compose build issues."
+        mkdir -p ~/.docker/cli-plugins
+        curl -sSLo ~/.docker/cli-plugins/docker-buildx "https://github.com/docker/buildx/releases/download/v0.17.1/buildx-v0.17.1.linux-amd64"
+        chmod +x ~/.docker/cli-plugins/docker-buildx
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,420 @@
+    "    # Run the query.\n",
+    "    # As additional input, include the descriptor data generated from our query image above.\n",
+    "    responses, images = db.query(q, [embedding])\n",
+    "    print(f\"{responses=}\")\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,420 @@
+    "\n",
+    "    print(f\"{ordered_images=}\")\n",
+    "\n",
+    "    if log_raw_output:\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,420 @@
+    "\n",
+    "\n",
+    "    \"\"\"\n",
+    "    Find similar images to the input embedding.\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "metadata": {},
+   "source": [
+    "## Query time!\n",
+    "### Find all the movies where Tom Hanks as been a part of "
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "professional = pd.json_normalize(response[1][\"FindEntity\"][\"entities\"])\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,442 @@
+    "            else:\n",
+    "                professional_details.append(f\"as {c['job']} in {c['department']}\")\n",
+    "\n",
+    "display(len(professional_details))\n",
+    "professional['details'] = professional_details\n",
+    "\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,442 @@
+    "query = \"\"\"\n",
+    "PREFIX wd: <http://www.wikidata.org/entity/>\n",
+    "PREFIX wdt: <http://www.wikidata.org/prop/direct/>\n",
+    "SELECT  ?title ?pop ?budget ?wTitle ?wmovie ?wTitle ?genreLabel ?loclabel\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries_nl.ipynb:
```diff
@@ -0,0 +1,205 @@
+   "outputs": [],
+   "source": [
+    "%pip install llama-index-llms-openai\n",
+    "%pip install --upgrade git+https://github.com/aperture-data/llama_index@kg_fixes#subdirectory=llama-index-integrations/graph_stores/llama-index-graph-stores-ApertureDB"
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/hello.ipynb:
```diff
@@ -32,6 +32,27 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Other notebooks\n",
+    "\n",
+    "In addition, the other notebooks provided in here depend on data on your instance.\n",
+    "| Notebook(s) | Data | Remarks | Related Workflow | \n",
+    "| --- | --- | --- | --- |\n",
```
Addressed in commit 2690f37.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_visualize.ipynb:
```diff
@@ -0,0 +1,311 @@
+    "                },\n",
+    "            }\n",
+    "        },\n",
+    "        {\n",
```
Addressed in commit 2690f37.

ad-claw000 committed:
Address review comments on PR 245

ad-claw000 commented on compose.sh:
```diff
@@ -51,5 +51,48 @@ DOCKERFILE_URL="${SOURCE_URL}/${SOURCE_PATH_REL}/Dockerfile"
+        echo "Workaround: Updating docker-buildx plugin to 0.17.1 to fix compose build issues."
+        mkdir -p ~/.docker/cli-plugins
+        OS=$(uname -s | tr "[:upper:]" "[:lower:]")
+        ARCH=$(uname -m)
```
Updated the script to respect DOCKER_CONFIG as suggested. See commit decb613.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_visualize.ipynb:
```diff
@@ -0,0 +1,312 @@
+    "        \"budget\": e['budget'],\n",
+    "        \"popularity\": e['popularity'],\n",
+    "        \"type\": \"MOVIE\"\n",
+    "    })\n",
```
Removed the unused `types` list and used `e.get('_class', ...)` consistently for types. See commit decb613.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,439 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%pip install --quiet mlcroissant pandas dotenv"
```
Removed unused dotenv import from pip install as suggested. See commit decb613.

ad-claw000 commented on apps/jupyterlab/app/notebooks/hello.ipynb:
```diff
@@ -32,6 +32,27 @@
+    "| ocr.ipynb | Documents scanned as images | This NB lets you interact with information extracted from text extracted from images | [OCR](https://docs.aperturedata.io/workflows/ocr_extraction) |\n",
+    "| rag.ipynb | A crawled website with segmentations and embeddings | | [Website Chatbot Workflow](https://docs.aperturedata.io/workflows/crawl_to_rag) |\n",
+    "| sql.ipynb | Interacts with a postgres server | Can be used with sql server workflow and Aperturedb | [SQL server](https://docs.aperturedata.io/workflows/sql_server) |\n",
+    "| movie_db | Notebooks to query a knowledge graph | Needs to have tmdb ingested into the Database | [Movie DB](movie_db/README.md)"
```
Added missing closing pipe for table formatting. See commit decb613.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries_nl.ipynb:
```diff
@@ -0,0 +1,205 @@
+    "    \"Movie\": [\"HasCast\", \"HasCrew\", \"HasGenre\", \"HasKeyword\"],\n",
+    "    \"Genre\": [\"HasCast\", \"HasCrew\", \"HasGenre\", \"HasKeyword\"],\n",
+    "    \"Keyword\": [\"HasCast\", \"HasCrew\", \"HasGenre\", \"HasKeyword\"],\n",
+    "}\n"
```
Removed unused variables `validation_schema`, `storage_context`, and `Settings` to keep the demo focused. See commit decb613.

ad-claw000 committed:
fix: address review comments on PR 245

ad-claw000 commented on compose.sh:
```diff
@@ -51,5 +51,49 @@ DOCKERFILE_URL="${SOURCE_URL}/${SOURCE_PATH_REL}/Dockerfile"
 echo "Description: ${DESCRIPTION}"
 
+# Workaround for docker compose build failing on older buildx versions
+if [[ " $* " == *" build "* ]]; then
```
Added a check for `--build` to ensure the workaround triggers for commands like `docker compose up --build`. See commit 2f9def4.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries_nl.ipynb:
```diff
@@ -0,0 +1,196 @@
+    "    possible_entities=entities,\n",
+    "    possible_relations=relations,\n",
+    "    possible_entity_props=[\"name\", \"budget\"],\n",
+    ")\n",
```
Replaced `Literal` with a concrete list of strings for `possible_entities` and `possible_relations` to prevent runtime errors, as suggested. See commit 2f9def4.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,439 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "query = \"\"\"\n",
+    "PREFIX wd: <http://www.wikidata.org/entity/>\n",
+    "PREFIX wdt: <http://www.wikidata.org/prop/direct/>\n",
+    "SELECT  ?title ?pop ?budget ?wmovie ?wTitle ?genreLabel ?loclabel\n",
+    "WHERE {\n",
+    "  ?m c:HasCast [p:name \"Tom Hanks\"] , [p:name \"Meg Ryan\"] ;\n",
+    "    p:title ?title ;\n",
+    "    p:popularity ?pop ;\n",
+    "    p:budget ?budget .\n",
+    "  ?m c:HasGenre ?genre.\n",
+    "  ?genre p:name ?genreLabel.\n",
+    "    BIND(strlang(?title, \"en\") AS ?stitle)\n",
+    "    SERVICE <https://query.wikidata.org/sparql> {\n",
+    "      ?wmovie wdt:P31 wd:Q11424.\n",
+    "      ?wmovie rdfs:label ?wTitle.\n",
+    "      ?wmovie wdt:P840 ?location.\n",
+    "      ?location rdfs:label ?loclabel.\n",
+    "      ?wmovie rdfs:label ?stitle.\n",
+    "      FILTER(lang(?wTitle) = \"en\")\n",
```
Added the missing `rdfs` prefix declaration to the SPARQL query to ensure it is portable. See commit 2f9def4.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_visualize.ipynb:
```diff
@@ -0,0 +1,311 @@
+    "for src in response[0]['FindEntity']['entities']:\n",
+    "    src_id = src['_uniqueid']\n",
+    "    for i, sr in enumerate(response[1:]):\n",
+    "        for e in sr['FindEntity']['entities'][src_id]:\n",
```
Replaced direct indexing with `.get(src_id, [])` to prevent `KeyError` when a source ID has no connected results. See commit 2f9def4.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,425 @@
+    "import pandas as pd\n",
+    "\n",
+    "def find_movie_posters(db: Connector,\n",
+    "    display_input:bool = True,\n",
+    "    constraints: Constraints = None,\n",
+    "    search_set_name: str = None,\n",
+    "    embedding: bytes = None,\n",
+    "    k_neighbors: int = 10,\n",
+    "    output_limit:int = 10,\n",
+    "    log_raw_output:bool = False) -> Images:\n",
+    "    \"\"\"\n",
```
Removed the unused `display_input` parameter from the function signature. See commit 2f9def4.

ad-claw000 committed:
fix: correctly update wf_ocr_done for processed images

luisremis commented on compose.sh:
```diff
@@ -51,5 +51,49 @@ DOCKERFILE_URL="${SOURCE_URL}/${SOURCE_PATH_REL}/Dockerfile"
 echo "Description: ${DESCRIPTION}"
 
+# Workaround for docker compose build failing on older buildx versions
+if [[ " $* " == *" build "* ]] || [[ " $* " == *" --build "* ]]; then
```
remove this workaround

ad-claw000 replied:
Removed the buildx workaround block entirely as requested. See commit 753e645.

ad-claw000 committed:
chore: remove buildx workaround from compose.sh

ad-claw000 committed:
Address review comments on demo notebooks

- Added OPENAI_API_KEY to README prerequisites
- Batched FindConnection queries in tmdb_queries to avoid O(N*M) separate requests
- Removed redundant StreamHandler in tmdb_queries_nl
- Gated constraint printing behind log_raw_output in tmdb_vector_search

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/README.md:
```diff
@@ -0,0 +1,16 @@
+
+- [Dataset Ingestion (Movies)](https://github.com/aperture-data/workflows/tree/main/apps/dataset-ingestion-movies) workflow must be executed to populate the database with the schema, cast, and movie data, including embeddings for poster images and taglines.
+- A running ApertureDB instance with Python client configured.
+
+## Notebooks
+
+- `tmdb_queries.ipynb`: Demonstrates how to perform property graph queries using the Python API.
+- `tmdb_queries_nl.ipynb`: Demonstrates how to perform natural language queries on the knowledge graph using LlamaIndex integration.
```
Added `OPENAI_API_KEY` to the prerequisites list in README.md. See commit 5118b58.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,440 @@
+    "                \"constraints\": {\n",
+    "                    \"_uniqueid\": [\"==\", src]\n",
+    "                },\n",
+    "                \"results\": {\n",
+    "                    \"all_properties\": True\n",
+    "                }\n",
+    "            }\n",
+    "        },\n",
+    "        {\n",
+    "            \"FindEntity\": {\n",
+    "                \"_ref\": 2,\n",
+    "                \"is_connected_to\": {\n",
+    "                    \"ref\": 1\n",
+    "                },\n",
+    "                \"with_class\": \"Movie\",\n",
+    "                \"constraints\": {\n",
+    "                    \"_uniqueid\": [\"==\", dst]\n",
+    "                },\n",
+    "                \"results\": {\n",
+    "                    \"all_properties\": True\n",
+    "                }\n",
+    "            }\n",
+    "        },{\n",
+    "            \"FindConnection\": {\n",
+    "                \"src\": 2,\n",
+    "                \"dst\": 1,\n",
+    "                \"results\": {\n",
+    "                    \"all_properties\": True\n",
+    "                }\n",
+    "            }\n",
+    "        }]\n",
+    "        _, responsec, _ = execute_query(client, q)\n",
+    "\n",
+    "        if responsec[2][\"FindConnection\"][\"returned\"] > 0:\n",
+    "            c = responsec[2][\"FindConnection\"][\"connections\"][0]\n",
+    "            if \"character\" in c:\n",
+    "                professional_details[dst] = f\"as character: {c['character']}\"\n",
+    "            else:\n",
+    "                professional_details[dst] = f\"as {c['job']} in {c['department']}\"\n",
+    "\n",
```
Refactored the nested loop to batch all `FindEntity` and `FindConnection` calls into a single query to avoid the O(N*M) HTTP overhead. See commit 5118b58.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries_nl.ipynb:
```diff
@@ -0,0 +1,195 @@
+    "import sys\n",
+    "\n",
+    "logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)\n",
+    "logging.getLogger().addHandler(logging.StreamHandler(stream=sys.stdout))\n",
```
Removed the redundant `StreamHandler` as suggested to avoid duplicate logging. See commit 5118b58.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,423 @@
+    "        }\n",
+    "    }\n",
+    "    if constraints is not None:\n",
+    "        print(constraints.constraints)\n",
```
Gated the constraint printing behind the `log_raw_output` parameter to keep notebook output clean by default. See commit 5118b58.

ad-claw000 commented on apps/ocr-extraction/app/image_ocr.py:
```diff
@@ -194,6 +186,27 @@ def response_handler(self, query, blobs, response, r_blobs):
+        done_query.append({
+            "FindImage": {
+                "_ref": 1,
+                "constraints": {
```
Updated the PR description to explicitly mention the OCR extraction pipeline optimization. See commit 5118b58 for the PR description update.

ad-claw000 committed:
fix: avoid empty uniqueids array in UpdateImage query

ad-claw000 committed:
merge: resolve conflicts with main

ad-claw000 committed:
Address review comments: remove unused imports, packages, and variables

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_vector_search.ipynb:
```diff
@@ -0,0 +1,425 @@
+   "outputs": [],
+   "source": [
+    "from IPython.display import display\n",
+    "import os\n",
```
Removed unused `import os` as suggested. See commit da2a6d6.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries.ipynb:
```diff
@@ -0,0 +1,441 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%pip install --quiet mlcroissant pandas"
```
Removed `mlcroissant` from the pip install list. See commit da2a6d6.

ad-claw000 commented on apps/jupyterlab/app/notebooks/movie_db/tmdb_queries_nl.ipynb:
```diff
@@ -0,0 +1,194 @@
+    "    ]\n",
+    "    for qs, is_correct in query_str:\n",
+    "        response = query_engine.query(qs)\n",
+    "        print(f\"{qs=}\")\n",
```
Dropped the `is_correct` flag by using `_` in the loop unpacking to avoid unused variable warning. See commit da2a6d6.

ad-claw000 committed:
fix(ocr): assert matching lengths and combine update query

ad-claw000 commented on apps/ocr-extraction/app/image_ocr.py:
```diff
@@ -194,6 +189,27 @@ def response_handler(self, query, blobs, response, r_blobs):
+                    "_uniqueid": ["in", uniqueids]
+                },
+            }
+        })
```
Added an assertion to ensure that `len(uniqueids) == len(r_blobs)` to fail fast if they diverge. See commit 2bce49e.

ad-claw000 commented on apps/ocr-extraction/app/image_ocr.py:
```diff
@@ -194,6 +189,27 @@ def response_handler(self, query, blobs, response, r_blobs):
+        })
+
+        status, r, _ = self.pool.execute_query(done_query)
+        assert status == 0, f"Query failed: {r}"
```
Combined the FindImage and UpdateImage steps into the same `desc_query` to ensure atomic execution. See commit 2bce49e.
